src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala

```diff
@@ -0,0 +1,216 @@
+            new BufferedReader(new StringReader(response.textBody))
+          )
+        )
+      case 304 =>
+        None
+      case _ =>
+        // we got an http error so we propagate it
+        log.error("HTTP Status {} in HTTP Source request: {}", response.status, response.asString)
+        throw HTTPException(Some(response.asString), response)
+    }
+  }
+
+  private def validateBodyLength(response: Response): Unit = {
+    val optContentLengthHeader = response.headers
+      .find(h => CONTENT_LENGTH.equalsIgnoreCase(h._1))
+      .map(h => h._2)
+      .map(l => l.toInt)
+    if (optContentLengthHeader.isEmpty) {
+      log.warn(s"Response doesn't contain $CONTENT_LENGTH header, only contained the following headers: ${response.headers.keySet}. Skipping validation...")
+      return
+    }
+
+    val contentLengthHeader = optContentLengthHeader.get
+    val bodyLength = response.asBytes.length
+    log.info(s"Validating body length ($bodyLength bytes) received from $uri against Content-Length header ($contentLengthHeader bytes) claimed in response")
+    try {
+      val reader = new BufferedReader(new StringReader(response.textBody))
+      val aclLineCount = reader.lines.count
+      log.info("There were {} lines in the response received from {}", aclLineCount, uri)
+      reader.close()
+    }
+    catch {
+      case e: Exception => log.warn(s"Failed to compute number of lines in the response received from $uri", e)
+    }
+    if (contentLengthHeader != bodyLength) {
+      val errorMessage = s"Body length ($bodyLength bytes) inconsistent with Content-Length header ($contentLengthHeader bytes)"
+      log.error(errorMessage)
+      throw HTTPException(Some(errorMessage), response)
+    }
+  }
+
+  private def getIdToken(): String = {
+    if (tokenCache == null
+      || Date.from(Instant.now()).after(tokenCache.getIdToken.getExpirationTime)) {
+      try {
+        log.info("Getting IdToken from Google Cloud")
+        tokenCache = getJwtFromGoogleIam
+      } catch {
+        case e: Exception =>
+          throw new RuntimeException("Couldn't get IdToken", e)
+      }
+    }
+    tokenCache.getRequestMetadata().get("Authorization").get(0) // Bearer Token
+  }
+
+  import java.io.IOException
+
+  @throws[IOException]
+  private def getJwtFromGoogleIam: IdTokenCredentials = {
+    val credentials = if (serviceAccountKey == null || serviceAccountKey.isEmpty) {
+      log.info("Getting Google Default Application Credentials...")
+      GoogleCredentials.getApplicationDefault
+    } else {
+      log.info("Getting Google Credentials from Key File...")
+      GoogleCredentials
+        .fromStream(new ByteArrayInputStream(this.serviceAccountKey.getBytes))
+        .createScoped(Collections.singleton(IAM_SCOPE))
+    }
+
+    credentials match {
+      case credentials: IdTokenProvider =>
+        IdTokenCredentials.newBuilder
+          .setIdTokenProvider(credentials.asInstanceOf[ServiceAccountCredentials])
```

Comment:
why the cast? seems not necessary

Reply:
Here we need to be sure to have an instance of `ServiceAccountCredentials` otherwise we can't use a credential key file

Reply:
so why not: `case credentials: ServiceAccountCredentials` ?

Reply:
Sure! I think I need more practice in Scala 🤦🏻‍♂️